File: iOS/Inventory/DittoManager.swift
//
//  DittoManager.swift
//  Inventory
//
//  Created by Shunsuke Kondo on 2023/01/19.
//  Copyright © 2023 Ditto. All rights reserved.
//

import Combine
import DittoSwift
import Foundation

public enum SyncEvent {
    case initial
    case update
}

// MARK: - Class Implementation
final class DittoManager {

    // MARK: - Models for Views
    struct Models {
        var items = [ItemDittoModel]()
    }
    
    private(set) var models = Models()

    // MARK: - Singleton object
    static let shared = DittoManager()

    // MARK: - Private Ditto properties
    private var ditto: Ditto!

    private var subscription: DittoSyncSubscription?
    private var storeObserver: DittoStoreObserver?

    // MARK: - Combine Subjects (to be observed from outside of this class)
    let itemsUpdated = PassthroughSubject<
        (indices: IndexSet, event: SyncEvent), Never
    >()

    // constructor is private because this is a singleton class
    private init() {
        startDitto()
    }

    /// Performs cleanup of Ditto resources
    ///
    /// This method handles the graceful shutdown of Ditto components by:
    /// - Cancelling any active subscriptions
    /// - Cancelling store observers
    /// - Stopping the Ditto sync process
    ///
    deinit {
        if ditto.isSyncActive {
            ditto.stopSync()
        }
        subscription?.cancel()
        subscription = nil
        storeObserver?.cancel()
        storeObserver = nil
    }

    private func startDitto() {
        DittoLogger.minimumLogLevel = .debug

        do {
            // Initialize Ditto
            // https://docs.ditto.live/sdk/latest/install-guides/swift#integrating-and-initializing-sync
            ditto = Ditto(
                identity:
                    .onlinePlayground(
                        appID: Env.DITTO_APP_ID,
                        token: Env.DITTO_PLAYGROUND_TOKEN,
                        enableDittoCloudSync: false,
                        customAuthURL: URL(string: Env.DITTO_AUTH_URL)
                    )
            )

            // Configure custom WebSocket endpoint for cloud sync
            ditto.updateTransportConfig { transportConfig in
                transportConfig.connect.webSocketURLs.insert(Env.DITTO_WEBSOCKET_URL)
            }

            // Disable sync with V3 Ditto
            try ditto.disableSyncWithV3()
            Task {
                // disable strict mode - allows for DQL with counters and objects as CRDT maps, must be called before startSync
                // https://docs.ditto.live/dql/strict-mode
                try await ditto.store.execute(
                    query: "ALTER SYSTEM SET DQL_STRICT_MODE = false"
                )
                try ditto.startSync()
            }
        } catch {
            let dittoErr = (error as? DittoSwiftError)?.errorDescription
            assertionFailure(dittoErr ?? error.localizedDescription)
        }
    }

    var dittoInfoView: DittoInfoViewController {
        DittoInfoViewFactory.create(ditto: ditto)
    }
}

// MARK: - Upsert Methods

extension DittoManager {

    /// Subscribes to all inventory items in the "inventories" collection.
    ///
    /// This method registers a subscription to the "inventories" collection using a DQL query.
    /// It also sets up an observer to monitor changes in the database, calculating the differences
    /// between syncs using a DittoDiffer. The observer sends updates to the `itemsUpdated` subject
    /// based on the type of change (initial load or update).
    ///
    /// - Note: This method should be called to start monitoring inventory items for changes.
    func subscribeAllInventoryItems() {
        let query = "SELECT * FROM inventories"
        do {
            // Create Subscription
            // https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions
            self.subscription = try ditto.sync.registerSubscription(
                query: query
            )

            // DittoDiffer - used to calculate the delta changes between syncs
            // https://docs.ditto.live/sdk/latest/crud/read#diffing-results
            let dittoDiffer = DittoDiffer()

            // Register Observer to see changes in the database from sync
            // https://docs.ditto.live/sdk/latest/crud/observing-data-changes
            storeObserver = try ditto.store.registerObserver(query: query) {
                [weak self] results in
                
                do {
                    let decoder = JSONDecoder()
                    let allItems = try results.items.compactMap{ try decoder.decode(ItemDittoModel.self, from: $0.jsonData()) }
                    self?.models.items = allItems
                    let diff = dittoDiffer.diff(results.items)
                    
                    // NOTE:  if you are curious on why we don't handle deletions - the app code
                    // currently does not allow deleting of inventory items, so there is no reason to handle
                    // checking the count of deletions.
                    
                    // if the insertions count is greater than zero and others are empty
                    // assume initial load
                    if diff.insertions.count > 0 && diff.deletions.isEmpty
                        && diff.updates.isEmpty
                    {
                        let event = SyncEvent.initial
                        self?.itemsUpdated.send(
                            (indices: diff.insertions, event: event)
                        )
                        
                    } else {
                        let event = SyncEvent.update
                        self?.itemsUpdated.send(
                            (indices: diff.updates, event: event)
                        )
                    }
                } catch {
                    print("Error: \(error)")
                }
            }
        } catch {
            print("Error: \(error)")
        }
    }

    /// Prepopulates the "inventories" collection with new items if they do not already exist.
    ///
    /// This method executes a DQL query to insert new documents into the "inventories" collection.
    /// It uses the "INSERT INTO inventories INTIAL DOCUMENTS" statement to ensure that the documents
    /// are only inserted if they do not already exist, matching the behavior of `.insertDefaultIfAbsent`.
    ///
    /// - Parameter itemIds: An array of integers representing the IDs of the items to be inserted.
    func prepopulateItemsIfAbsent(itemIds: [Int]) {

        // CREATE new items using the INSERT INTO xxx INTIAL statement
        // https://docs.ditto.live/dql/insert#insert-with-initial-documents
        let query = "INSERT INTO inventories INITIAL DOCUMENTS (:item)"

        Task {
            // Create a transaction to run inserts into with DQL - this is the equivalent to scoped transaction using store.write
            // https://docs.ditto.live/sdk/latest/crud/transactions
            do {
                try await ditto.store.transaction { transaction in
                    for itemId in itemIds {
                        do {
                            try await transaction.execute(
                                query: query,
                                arguments: [
                                    "item":
                                        [
                                            "_id": itemId,
                                            "counter": 0.0
                                        ]
                                ]
                            )
                        } catch {
                            let dittoErr = (error as? DittoSwiftError)?
                                .errorDescription
                            assertionFailure(
                                dittoErr ?? error.localizedDescription
                            )
                            return .rollback
                        }
                    }
                    return .commit
                }
            } catch {
                let dittoErr = (error as? DittoSwiftError)?
                    .errorDescription
                assertionFailure(dittoErr ?? error.localizedDescription)
            }
        }
    }

    func incrementCounterFor(id: Int) {
        // UPDATE Counter using DQL PN_INCREMENT function
        // TODO insert URL to documentation link
        let query =
            "UPDATE inventories APPLY counter PN_INCREMENT BY 1.0 WHERE _id = :id"
        Task {
            do {
                try await ditto.store.execute(query: query, arguments: ["id": id])
            } catch {
                let dittoErr = (error as? DittoSwiftError)?
                    .errorDescription
                assertionFailure(dittoErr ?? error.localizedDescription)
            }
        }
    }

    func decrementCounterFor(id: Int) {
        // UPDATE Counter using DQL PN_INCREMENT function
        // TODO insert URL to documentation link
        let query =
            "UPDATE inventories APPLY counter PN_INCREMENT BY -1.0 WHERE _id = :id"
        Task {
            do {
                try await ditto.store.execute(query: query, arguments: ["id": id])
            } catch {
                let dittoErr = (error as? DittoSwiftError)?
                    .errorDescription
                assertionFailure(dittoErr ?? error.localizedDescription)
            }
        }
    }
}

File: Android/app/src/main/java/live/ditto/inventory/DittoManager.kt
package live.ditto.inventory

import android.content.Context
import android.util.Log
import live.ditto.*
import live.ditto.android.DefaultAndroidDittoDependencies

object DittoManager {
    /* Interfaces */
    interface ItemUpdateListener {
        fun setInitial(items: List<ItemModel>)
        fun updateCount(index: Int, count: Int)
    }

    /* Settable from outside */
    lateinit var itemUpdateListener: ItemUpdateListener

    /* Get-only properties */
    var ditto: Ditto? = null; private set

    /* Private properties */
    private var subscription: DittoSyncSubscription? = null
    private var observer: DittoStoreObserver? = null

    // Those values should be pasted in 'gradle.properties'. See the notion page for more details.
    private const val APP_ID = BuildConfig.DITTO_APP_ID
    private const val ONLINE_AUTH_TOKEN = BuildConfig.DITTO_PLAYGROUND_TOKEN
    private const val AUTH_URL = BuildConfig.DITTO_AUTH_URL
    private const val WEBSOCKET_URL = BuildConfig.DITTO_WEBSOCKET_URL

    /* Internal functions and properties */
    internal suspend fun startDitto(context: Context) {
        DittoLogger.minimumLogLevel = DittoLogLevel.DEBUG

        val dependencies = DefaultAndroidDittoDependencies(context)
        ditto = Ditto(dependencies, DittoIdentity.OnlinePlayground(dependencies, APP_ID, ONLINE_AUTH_TOKEN, false, AUTH_URL))

        try {
            ditto?.let {
                // Configure custom WebSocket endpoint for cloud sync
                it.updateTransportConfig { transportConfig ->
                    transportConfig.connect.websocketUrls.add(WEBSOCKET_URL)
                }

                // Disable sync with V3 Ditto
                it.disableSyncWithV3()

                // disable strict mode - allows for DQL with counters and objects as CRDT maps, must be called before startSync
                // https://docs.ditto.live/dql/strict-mode
                it.store.execute("ALTER SYSTEM SET DQL_STRICT_MODE = false")

                // start sync
                // https://docs.ditto.live/sdk/latest/sync/start-and-stop-sync
                it.startSync()
            }

        } catch (e: Throwable) {
            e.localizedMessage?.let { Log.e(e.message, it) }
        }

        observeItems()
        insertDefaultDataIfAbsent()
    }

    internal suspend fun increment(itemId: Int) {
        // UPDATE Counter using DQL PN_INCREMENT function
        // TODO insert URL to documentation link
        val query = "UPDATE inventories APPLY counter PN_INCREMENT BY 1.0 WHERE _id = :id"
        try {
            ditto?.store?.execute(query,
                mapOf("id" to itemId))
        } catch (e: Throwable){
            e.localizedMessage?.let { Log.e(e.message, it) }
        }
    }

    internal suspend fun decrement(itemId: Int) {
        // UPDATE Counter using DQL PN_INCREMENT function
        // TODO insert URL to documentation link
        val query = "UPDATE inventories APPLY counter PN_INCREMENT BY -1.0 WHERE _id = :id"
        try {
            ditto?.store?.execute(query,
                mapOf("id" to itemId))
        } catch (e: Throwable){
            e.localizedMessage?.let { Log.e(e.message, it) }
        }
    }

    internal val sdkVersion: String?
        get() = ditto?.sdkVersion


    /* Private functions and properties */
    private suspend fun insertDefaultDataIfAbsent() {
        // CREATE new items using the INSERT INTO xxx INTIAL statement
        // https://docs.ditto.live/dql/insert#insert-with-initial-documents
        val query = "INSERT INTO inventories INITIAL DOCUMENTS (:item)"

        // Create a transaction to run inserts into with DQL - this is the equivalent to scoped transaction using store.write
        // https://docs.ditto.live/sdk/latest/crud/transactions
        ditto?.store?.transaction {
             try {
                 for (viewItem in itemsForView) {
                     it.execute(query,
                         mapOf("item" to
                                 mapOf("_id" to viewItem.itemId,
                                     "counter" to 0.0)
                         )
                     )
                 }
             } catch (e: Throwable){
                 e.localizedMessage?.let { Log.e(e.message, it) }
                 DittoTransactionCompletionAction.Rollback
             }
            DittoTransactionCompletionAction.Commit
        }
    }

    private fun observeItems() {
        val query = "SELECT * FROM inventories"
        ditto?.let {
            // Create Subscription
            // https://docs.ditto.live/sdk/latest/sync/syncing-data#creating-subscriptions
            subscription = it.sync.registerSubscription(query)

            // DittoDiffer - used to calculate the delta changes between syncs
            // https://docs.ditto.live/sdk/latest/crud/read#diffing-results
            val dittoDiffer = DittoDiffer()

            // Register Observer to see changes in the database from sync
            // https://docs.ditto.live/sdk/latest/crud/observing-data-changes
            observer = it.store.registerObserver(query) { results ->
                val diff = dittoDiffer.diff(results.items)

                // NOTE:  if you are curious on why we don't handle deletions - the app code
                // currently does not allow deleting of inventory items, so there is no reason to handle
                // checking the count of deletions.

                // if the insertions count is greater than zero and others are empty
                // assume initial load
                if (diff.insertions.isNotEmpty() && diff.deletions.isEmpty() && diff.updates.isEmpty()) {
                   itemUpdateListener.setInitial(itemsForView.toMutableList())
                } else {
                    diff.updates.forEach { index ->
                        val doc = results.items[index].value
                        val count = doc["counter"] as Float
                        itemUpdateListener.updateCount(index, count.toInt())
                    }
                }
            }
        }
    }

    private val itemsForView = arrayOf(
        ItemModel(0, R.drawable.coke, "Coca-Cola", 2.50, "A Can of Coca-Cola"),
        ItemModel(1, R.drawable.drpepper, "Dr. Pepper", 2.50, "A Can of Dr. Pepper"),
        ItemModel(2,R.drawable.lays, "Lay's Classic", 3.99, "Original Classic Lay's Bag of Chips"),
        ItemModel(3, R.drawable.brownies, "Brownies", 6.50,"Brownies, Diet Sugar Free Version"),
        ItemModel(4, R.drawable.blt, "Classic BLT Egg", 2.50, "Contains Egg, Meats and Dairy")
    )
}